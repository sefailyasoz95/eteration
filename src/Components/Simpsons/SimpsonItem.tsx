import React, { useEffect, useState } from "react";
import { View, Text, StyleSheet, Image, TouchableOpacity, GestureResponderEvent, Alert } from "react-native";
import { imageUrlHelper } from "../../Utils/Helpers/imageUrlHelper";
import { deleteSimpson } from "../../Utils/Services/service";
type Props = {
	item: any;
	onPress: (event: GestureResponderEvent) => void;
	onDelete: (event: GestureResponderEvent) => void;
};

const SimpsonItem: React.FC<Props> = ({ item, onPress, onDelete }) => {
	const [isValidURL, setIsValidURL] = useState(false);
	useEffect(() => {
		setIsValidURL(imageUrlHelper(item.avatar));
	}, []);

	return (
		<TouchableOpacity style={styles.itemContainer} onPress={() => onPress(item)}>
			<View style={styles.imageAndName}>
				{isValidURL ? (
					<Image source={{ uri: item.avatar }} width={50} height={50} style={styles.image} />
				) : (
					<Text style={styles.noImage}>â›”</Text>
				)}
				<Text style={styles.name}>{item.name}</Text>
			</View>
			<TouchableOpacity onPress={() => onDelete(item)}>
				<Text style={styles.deleteButton}>{`x`} </Text>
			</TouchableOpacity>
		</TouchableOpacity>
	);
};

const styles = StyleSheet.create({
	itemContainer: {
		padding: 10,
		marginVertical: 5,
		flexDirection: "row",
		alignItems: "center",
		justifyContent: "space-between",
		marginHorizontal: 5,
		backgroundColor: "#fff",
		elevation: 5,
		shadowColor: "black",
		shadowOffset: {
			width: 0,
			height: 1,
		},
		shadowOpacity: 0.3,
		shadowRadius: 3,
		borderRadius: 10,
	},
	image: {
		height: 50,
		width: 50,
		resizeMode: "contain",
	},
	noImage: {
		padding: 10,
	},
	name: {
		letterSpacing: 1,
		marginHorizontal: 10,
	},
	imageAndName: {
		flexDirection: "row",
		alignItems: "center",
		justifyContent: "flex-start",
	},
	deleteButton: {
		fontSize: 25,
		fontWeight: "bold",
		color: "#ff2222",
		padding: 10,
	},
});

export default SimpsonItem;
