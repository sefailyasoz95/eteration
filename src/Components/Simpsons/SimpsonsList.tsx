import { NativeStackNavigationProp } from "@react-navigation/native-stack";
import React, { useCallback, useState } from "react";
import { Alert, FlatList, RefreshControl, StyleSheet } from "react-native";
import { AppStackParamList } from "../../StackParamLists/AppStackParamList";
import { deleteSimpson } from "../../Utils/Services/service";
import SimpsonItem from "./SimpsonItem";

type Props = {
	dataSource: Array<Object>;
	navigation: NativeStackNavigationProp<AppStackParamList, "List">;
	refreshData: Function;
	testID: string;
};

const SimpsonsList: React.FC<Props> = ({ dataSource, navigation, refreshData, testID }) => {
	const showSimpsonDetail = (simpson: any) => {
		navigation.navigate("Detail", { simpson });
	};
	const [refreshing, setRefreshing] = useState(false);
	const onRefresh = useCallback(() => {
		setRefreshing(true);
		refreshData();
		setRefreshing(false);
	}, []);
	const handleDelete = (item: any) => {
		Alert.alert("Hey there!", `Are you sure to delete ${item.name}?`, [
			{
				text: "Cancel",
				onPress: () => null,
				style: "cancel",
			},
			{
				text: "Yes",
				onPress: async () => {
					await deleteSimpson(item.id);
					refreshData();
				},
			},
		]);
	};
	return (
		<FlatList
			data={dataSource}
			testID={testID || "flatlist"}
			contentContainerStyle={styles.containerStyle}
			style={styles.flatList}
			refreshControl={<RefreshControl refreshing={refreshing} onRefresh={onRefresh} />}
			renderItem={({ item, index }) => (
				<SimpsonItem key={index} testID='simpsonItem' item={item} onPress={showSimpsonDetail} onDelete={handleDelete} />
			)}
		/>
	);
};

const styles = StyleSheet.create({
	containerStyle: {},
	flatList: {
		maxHeight: "80%",
		marginTop: "5%",
		width: "95%",
	},
});

export default SimpsonsList;
