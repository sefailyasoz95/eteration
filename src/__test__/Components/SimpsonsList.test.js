import React from "react";
import SimpsonsList from "../../Components/Simpsons/SimpsonsList";
import { shallow, configure } from "enzyme";
import toJson from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";
import { render } from "@testing-library/react-native";
import SimpsonItem from "../../Components/Simpsons/SimpsonItem";

configure({ adapter: new Adapter() });

const simpsons = [
	{ id: 1, name: "Merge Simpson", about: "something about Merge", job: "Merge's job", avatar: "https://image.com" },
	{ id: 2, name: "Homer Simpson", about: "something about Homer", job: "Homer's job", avatar: "https://image.com" },
];

describe("Simpsons List Component", () => {
	it("should render without issues", () => {
		const component = shallow(<SimpsonsList dataSource={simpsons} />);

		expect(component.length).toBe(1);
		expect(toJson(component)).toMatchSnapshot();
	});

	it("should render all item in simpsonlist", () => {
		const wrapper = render(<SimpsonsList dataSource={simpsons} />);

		// check if the flatlist item renders correctly
		expect(wrapper.getByTestId("flatlist")).toBeTruthy();
		expect(wrapper.UNSAFE_getAllByType(SimpsonItem).length).toBe(simpsons.length);
	});

	it("should render correct names", () => {
		const wrapper = shallow(<SimpsonsList dataSource={simpsons} />);

		wrapper.find({ testID: "simpsonItem" }).forEach((node, index) => {
			expect(node.props().children).toBe(simpsons[index].name);
		});
	});
});
