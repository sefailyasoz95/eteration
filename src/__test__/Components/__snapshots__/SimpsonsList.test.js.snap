// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simpsons List Component should render without issues 1`] = `
<FlatList
  contentContainerStyle={Object {}}
  data={
    Array [
      Object {
        "about": "something about Merge",
        "avatar": "https://image.com",
        "id": 1,
        "job": "Merge's job",
        "name": "Merge Simpson",
      },
      Object {
        "about": "something about Homer",
        "avatar": "https://image.com",
        "id": 2,
        "job": "Homer's job",
        "name": "Homer Simpson",
      },
    ]
  }
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onEndReachedThreshold={2}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  style={
    Object {
      "marginTop": "5%",
      "maxHeight": "80%",
      "width": "95%",
    }
  }
  testID="flatlist"
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;
